lib/ansible/module_utils/csharp/Ansible.Become.cs

```diff
@@ -1,690 +1,870 @@
         }
 
-        private static uint GetProcessExitCode(IntPtr processHandle)
+        private static SecurityIdentifier GetTokenUserSID(SafeNativeHandle hToken)
         {
-            new NativeWaitHandle(processHandle).WaitOne();
-            uint exitCode;
-            if (!GetExitCodeProcess(processHandle, out exitCode))
-                throw new Win32Exception("Error getting process exit code");
-            return exitCode;
+            using (SafeMemoryBuffer tokenInfo = GetTokenInformation(hToken, NativeHelpers.TokenInformationClass.TokenUser))
+            {
+                NativeHelpers.TOKEN_USER tokenUser = (NativeHelpers.TOKEN_USER)Marshal.PtrToStructure(tokenInfo.DangerousGetHandle(),
+                    typeof(NativeHelpers.TOKEN_USER));
+                return new SecurityIdentifier(tokenUser.User.Sid);
+            }
         }
 
-        private static IntPtr GetElevatedToken(IntPtr hToken)
+        private static NativeHelpers.SECURITY_LOGON_TYPE GetTokenLogonType(SafeNativeHandle hToken)
         {
-            uint requestedLength;
-
-            IntPtr pTokenInfo = Marshal.AllocHGlobal(sizeof(int));
+            UInt64 tokenLuidId;
+            using (SafeMemoryBuffer tokenInfo = GetTokenInformation(hToken, NativeHelpers.TokenInformationClass.TokenStatistics))
+            {
+                NativeHelpers.TOKEN_STATISTICS stats = (NativeHelpers.TOKEN_STATISTICS)Marshal.PtrToStructure(
+                    tokenInfo.DangerousGetHandle(), typeof(NativeHelpers.TOKEN_STATISTICS));
+                tokenLuidId = (UInt64)stats.AuthenticationId;
+            }
 
-            try
+            // Default to Network, if we weren't able to get the actual type treat it as an error and assume
+            // we don't want to run a process with the token
+            NativeHelpers.SECURITY_LOGON_TYPE logonType = NativeHelpers.SECURITY_LOGON_TYPE.Network;
+            UInt32 sessionCount;
+            SafeLsaMemoryBuffer sessionPtr;
+            UInt32 res = NativeMethods.LsaEnumerateLogonSessions(out sessionCount, out sessionPtr);
+            if (res != 0)
+                throw new Win32Exception((int)NativeMethods.LsaNtStatusToWinError(res), "LsaEnumerateLogonSession() failed");
+            using (sessionPtr)
             {
-                if (!GetTokenInformation(hToken, TokenInformationClass.TokenElevationType, pTokenInfo, sizeof(int), out requestedLength))
-                    throw new Win32Exception("Unable to get TokenElevationType");
+                IntPtr currentSession = sessionPtr.DangerousGetHandle();
+                for (UInt32 i = 0; i < sessionCount; i++)
+                {
+                    SafeLsaMemoryBuffer sessionDataPtr;
+                    res = NativeMethods.LsaGetLogonSessionData(currentSession, out sessionDataPtr);
+                    if (res != 0)
+                    {
+                        currentSession = IntPtr.Add(currentSession, Marshal.SizeOf(typeof(NativeHelpers.LUID)));
```

Comment:
Rather than duplicating this for the skip case, maybe a try/finally inside the loop (just put the one at the bottom of the loop in the finally)? Or maybe do the pointer math directly in the for loop. Not sure if either of those ends up with easier-read code, so your call.

Reply:
Done